subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/asm/ClassDependenciesVisitor.java

```diff
@@ -73,6 +75,16 @@ public static ClassAnalysis analyze(String className, ClassReader reader, String
     public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
         isAnnotationType = isAnnotationType(interfaces);
         Set<String> types = isAccessible(access) ? accessibleTypes : privateTypes;
+        if (signature != null) {
+            SignatureReader signatureReader = new SignatureReader(signature);
+            signatureReader.accept(new SignatureVisitor(API) {
+                @Override
+                public void visitClassType(String className) {
+                    Type type = Type.getObjectType(className);
+                    maybeAddDependentType(types, type);
```

Comment:
I wonder if generics are part of ABI or we could always add them to `privateTypes`?